File: day2.nim
import std/[sequtils, strscans]

var
    depth = 0
    hpos = 0
    aim = 0

let data = "input2.txt".lines.toSeq

proc part1 =
    var cmd: string
    var delta: int

    for line in data:
        if scanf(line, "$w $i", cmd, delta):
            case cmd
            of "forward": hpos += delta
            of "up": depth -= delta
            of "down": depth += delta
            else: doAssert(false, "Unknown command: " & cmd)

    echo depth * hpos

proc part2 =
    var cmd: string
    var delta: int
    for line in data:
        if scanf(line, "$w $i", cmd, delta):
            case cmd
            of "forward":
                hpos += delta
                depth += aim * delta
            of "up":
                aim -= delta
            of "down":
                aim += delta
            else: doAssert(false, "Unknown command: " & cmd)
    echo depth * hpos

part2()

File: cs/Day2.cs
using System;
using System.IO;
using System.Collections.Generic;

class Day2 {
  public void solve() {
    int pos = 0;
    int depth = 0;
    int aim = 0;

    var data = File.ReadAllText("../input2.txt").Trim().Split('\n').Select(line => {
        var parts = line.Split(' ');
        return (parts[0], int.Parse(parts[1]));
    });

    foreach (var (cmd, delta) in data) {
      switch (cmd) {
        case "forward":
          pos += delta;
          depth += aim * delta;
          break;
        case "down":
          aim += delta;
          break;
        case "up":
          aim -= delta;
          break;
      };
    }

    Console.WriteLine(pos * depth);
  }
}
